src/proxy/outbound.rs

```diff
@@ -470,15 +470,26 @@ impl OutboundConnection {
             )
             .await?;
 
+        // TODO src_id may not be enough; should also check addresses/uid
+        let src_id = Some(source_workload.identity());
+        let from_waypoint = proxy::check_from_waypoint(
+            self.pi.state.clone(),
+            &mutable_us.workload,
+            src_id.as_ref(),
+        )
+        .await;
+
         // For case upstream server has enabled waypoint
-        match self
-            .pi
-            .state
-            .fetch_waypoint(&mutable_us.workload, workload_ip)
-            .await
-        {
-            Ok(None) => {} // workload doesn't have a waypoint; this is fine
-            Ok(Some(waypoint_us)) => {
+        match (
+            from_waypoint,
+            self.pi
+                .state
+                .fetch_waypoint(&mutable_us.workload, workload_ip)
+                .await,
+        ) {
+            (_, Ok(None)) => {} // workload doesn't have a waypoint; this is fine
+            (true, _) => {}     // we already traversed the waypoint
+            (false, Ok(Some(waypoint_us))) => {
```

Comment:
In this branch (and the one for to Service) we're using the port WDS GatewayAddress sends us. 

https://github.com/istio/ztunnel/blob/a17e43e3488a4567c4c84eaae0638bf09d045c4d/src/proxy/outbound.rs#L423

Right now, the CP happens to send us 15008 which happens to be what the ztunnel listens on.

Reply:
Is the CP going to be smart and send the zTunnel address, or do we need to be smart and use `pi.hbone_port`?

Reply:
Right now as per the protocol 15008 is "the" HBONE port and we assume that in other spots, I suspect.

For this I would suggest maybe splitting the difference and taking what CP sends us, but explicitly checking here that `waypoint_us.port == pi.hbone_port` and throwing an error if not. If we wish to intentionally change that behavior later we can.